src/permissions.js

```diff
@@ -214,5 +214,21 @@ export function getManifestOriginMatchPatterns() {
             }
         }
     }
+
+    // Manifest v3 requires host match patterns to be in an array under the host_permissions key.
+    if ("manifest_version" in manifest && manifest.manifest_version === 3) {
```

Comment:
Any reason not to if/else this with the above? or alternatively, combine with something like this:

```
const manifestPermissions = [ ];

("permissions" in manifest) &&
Array.isArray(manifest.permissions) &&
manifestPermissions.push(...manifest.permissions);

("host_permissions" in manifest) &&
Array.isArray(manifest.host_permissions) &&
manifestPermissions.push(...manifest.host_permissions);

for (const permission of manifestPermissions) {
...
```

Note, I also don't think it's necessary to check for the manifest version? But I understand wanting to have some indication for clarity (maybe in a comment).

Reply:
So the reason we check for manifest version is that in v2, match patterns went in `permissions` like:

```
permissions: [
  "tabs",
  "...",
  "<all_urls>"
]
```

Where in manifest v3, there's a new `host_permissions` to hold just match patterns.

So I think what we want is in mv2 to continue to look in `permissions`, and in mv3 use `host_permissions` instead.

I'll add a comment, does that all sound OK and correct @aaga ?

Reply:
This is covered in https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#host-permissions among other docs

Reply:
oh yes I understood that part, sorry I should have been more verbose with my comment.

my suggestion is mostly about structure not functionality. currently we have two identical for loops, just looping over different arrays. It makes sense to me to combine those two arrays and just have one loop. Additionally, for the manifest v3 case, the code still first loops over the non-host permissions, even though we know there will be nothing of interest there.

Reply:
Oh, I see, thanks.

Reply:
OK I updated it as you suggested, although it occurs to me that in MV3 there should never be any host match patterns in the permissions so it's a bit of wasted effort, but this list is likely to be very small in every imaginable case so I think over-optimizing isn't worth it vs. code clarity (n will always be small).

Also the MV2 bits of this code can go away when Firefox transitions.